int64[] row1
int64[] row2
int64[] row3